litestar/routes/http.py

```diff
@@ -78,14 +113,37 @@ async def handle(self, scope: HTTPScope, receive: Receive, send: Send) -> None:
             await route_handler.authorize_connection(connection=request)
 
         try:
-            response = await self._get_response_for_request(
-                scope=scope, request=request, route_handler=route_handler, parameter_model=parameter_model
-            )
-
-            await response(scope, receive, send)
+            if route_handler.has_sync_callable or isinstance(route_handler.fn, AsyncCallable):
+                # if it's a sync or to_thread function we can't actually cancel anything
+                # so we just await it directly
+                await self._handle_response_cycle(
+                    scope=scope,
+                    send=send,
+                    receive=receive,
+                    request=request,
+                    route_handler=route_handler,
+                    parameter_model=parameter_model,
+                )
+            else:
+                async with anyio.create_task_group() as tg:
+                    tg.start_soon(
+                        partial(
+                            self._handle_response_cycle,
+                            scope=scope,
+                            send=send,
+                            receive=receive,
+                            request=request,
+                            route_handler=route_handler,
+                            parameter_model=parameter_model,
+                            cancel_scope=tg.cancel_scope,
+                        ),
+                    )
+                    tg.start_soon(self._listen_for_disconnect, request, tg.cancel_scope)
```

Comment:
This is a more performant, but we can't really use it for reasons of anyio/trio compatibility.

```suggestion
tasks = [
    asyncio.create_task(
        self._handle_response_cycle(
            scope=scope,
            send=send,
            receive=receive,
            request=request,
            route_handler=route_handler,
            parameter_model=parameter_model,
        )
    ),
    asyncio.create_task(request._listen_for_disconnect()),
]
done, pending = await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)
done.pop().result()
if done:
    done.pop().result()

if pending:
    pending = pending.pop()
    pending.cancel()
    await pending
    pending.result()
```